cmd/hydrate/src/lib.rs

```diff
@@ -0,0 +1,212 @@
+
+    /// Path to the raw memory dump
+    file: PathBuf,
+}
+
+struct DryCore {
+    flash: HubrisFlashMap,
+    mem: BTreeMap<u32, Vec<u8>>,
+}
+
+// Helper macro to stub out functions
+macro_rules! unsupported{
+    ($fn_name:ident($($arg_name:ident: $arg_type:ty),*)) => {
+        unsupported!($fn_name($($arg_name: $arg_type),*) -> Result<()>);
+    };
+    ($fn_name:ident($($arg_name:ident: $arg_type:ty),*) -> $out:ty) => {
+        fn $fn_name(&mut self, $($arg_name: $arg_type),*) -> $out {
+            bail!(concat!(
+                "DryCore does not support ",
+                stringify!($fn_name)))
+        }
+    };
+}
+
+impl humility::core::Core for DryCore {
+    unsupported!(run());
+    unsupported!(halt());
+    unsupported!(step());
+    unsupported!(load(_path: &std::path::Path));
+    unsupported!(reset());
+    unsupported!(write_8(_addr: u32, _data: &[u8]));
+    unsupported!(op_done());
+    unsupported!(op_start());
+    unsupported!(read_reg(_reg: ARMRegister) -> Result<u32>);
+    unsupported!(init_swv());
+    unsupported!(read_swv() -> Result<Vec<u8>>);
+    unsupported!(write_reg(_reg: ARMRegister, _value: u32));
+    unsupported!(write_word_32(_addr: u32, _data: u32));
+    unsupported!(reset_and_halt(_dur: std::time::Duration));
+    unsupported!(wait_for_halt(_dur: std::time::Duration));
+
+    fn info(&self) -> (String, Option<String>) {
+        ("DryCore".to_owned(), None)
+    }
+    fn is_archive(&self) -> bool {
+        false
+    }
+    fn is_dump(&self) -> bool {
+        true // I guess?
+    }
+    fn is_net(&self) -> bool {
+        false
+    }
+    fn vid_pid(&self) -> Option<(u16, u16)> {
+        None
+    }
+
+    fn read_8(&mut self, addr: u32, data: &mut [u8]) -> Result<()> {
+        if self.flash.read(addr, data).is_some() {
+            return Ok(());
+        }
+
+        let Some((base, mem)) = self.mem.range(0..=addr).next_back() else {
+            bail!("addr {addr:#08x} is below memory range");
+        };
+        let offset = (addr - base) as usize;
+        let end = offset + data.len();
+        if end > mem.len() {
+            bail!("region is not large enough; {end:#x} > {:#x}", mem.len());
+        }
+        data.copy_from_slice(&mem[offset..end]);
+        Ok(())
+    }
+}
+
+fn run(context: &mut ExecutionContext) -> Result<()> {
+    let Subcommand::Other(subargs) = context.cli.cmd.as_ref().unwrap();
+    let subargs = Args::try_parse_from(subargs)?;
+    let f = std::fs::File::open(&subargs.file)?;
+    let mut z = zip::ZipArchive::new(f)?;
+
+    let mut s = String::new();
+    z.by_name("dump.json")?
+        .read_to_string(&mut s)
+        .with_context(|| "could not find `dump.json` in the ZIP archive")?;
+
+    #[derive(serde::Deserialize)]
+    struct DumpInfo {
+        format: u64,
+        task_index: u16,
+        crash_time: u64,
+        board_name: String,
+        git_commit: String,
+        archive_id: String,
+        fw_version: Option<String>,
+    }
```

Comment:
part of me wonders if it might be worth having a shared definition of this structure that could also be used by MGS instead of the [ad-hoc JSON here](https://github.com/oxidecomputer/management-gateway-service/pull/316/files#diff-eecb3109d05f85699dc42276b1a71b8ec4041db954cce5ba3a1496c386685bf0R1583-R1589), but... i don't really know what crate we would put that in that it would make sense to have as a dependency of both MGS and humility. so, maybe it's not worth the effort... :woman_shrugging:

Reply:
I think the dependency should probably go Humility → something in `management-gateway`, but agree that it's not worth the effort / coupling yet...